examples/image-search/build.cpp

```diff
@@ -0,0 +1,155 @@
+#include "clip.h"
+#include "usearch/index.hpp"
+
+#include <fstream>
+#include <filesystem>
+
+struct my_app_params {
+	int32_t n_threads {4};
+	std::string model = "../models/ggml-model-f16.bin";
+	std::vector<std::string> image_directories;
+};
+
+void my_print_help(int argc, char **argv, my_app_params &params) {
+	printf("Usage: %s [options] dir/with/pictures [more/dirs]\n", argv[0]);
+	printf("\nOptions:");
+	printf("  -h, --help: Show this message and exit\n");
+	printf("  -m <path>, --model <path>: path to model. Default: %s\n", params.model.c_str());
+	printf("  -t N, --threads N: Number of threads to use for inference. Default: %d\n", params.n_threads);
+}
+
+// returns success
+bool my_app_params_parse(int argc, char **argv, my_app_params &params) {
+	bool invalid_param = false;
+	for (int i = 1; i < argc; i++) {
+
+		std::string arg = argv[i];
+
+		if (arg == "-m" || arg == "--model") {
+			if (++i >= argc) {
+				invalid_param = true;
+				break;
+			}
+			params.model = argv[i];
+		} else if (arg == "-t" || arg == "--threads") {
+			if (++i >= argc) {
+				invalid_param = true;
+				break;
+			}
+			params.n_threads = std::stoi(argv[i]);
+		} else if (arg == "-h" || arg == "--help") {
+			my_print_help(argc, argv, params);
+			exit(0);
+		} else if (arg.starts_with('-')) {
+			if (i != 0) {
+				printf("%s: unrecognized argument: %s\n", __func__, arg.c_str());
+				return false;
+			}
+		} else {
+			// assume image directory
+			params.image_directories.push_back(argv[i]);
+		}
+	}
+
+	return !(invalid_param || params.image_directories.empty());
+}
+
+bool is_image_file_extension(std::string_view ext) {
+	if (ext == ".jpg") return true;
+	if (ext == ".JPG") return true;
+
+	if (ext == ".jpeg") return true;
+	if (ext == ".JPEG") return true;
+
+	if (ext == ".gif") return true;
+	if (ext == ".GIF") return true;
+
+	if (ext == ".png") return true;
+	if (ext == ".PNG") return true;
+
+	// TODO: add all the supported files for std_image
+
+	return false;
+}
+
+int main(int argc, char** argv) {
+	my_app_params params;
+	if (!my_app_params_parse(argc, argv, params)) {
+		my_print_help(argc, argv, params);
+		return 1;
+	}
+
+	auto clip_ctx = clip_model_load(params.model.c_str());
+	if (!clip_ctx) {
+		printf("%s: Unable  to load model from %s\n", __func__, params.model.c_str());
+		return 1;
+	}
+
+	std::vector<std::string> image_file_index;
+	unum::usearch::index_gt<unum::usearch::cos_gt<float>> embd_index;
+
+	const size_t vec_dim = clip_ctx->vision_model.hparams.projection_dim;
+
+	size_t label = 0;
+
+	// search for images in path and create embedding
+	for (const auto& base_dir : params.image_directories) {
+		fprintf(stdout, "%s: starting base dir scan of '%s'\n", __func__, base_dir.c_str());
+
+		for (auto const& dir_entry : std::filesystem::recursive_directory_iterator(base_dir)) {
+			if (!dir_entry.is_regular_file()) {
```

Comment:
Do we need C++17 just for this? According to their header, USearch requires only C++11. And implementing this recursive directory walk should not be that hard in C++11 as well.

Reply:
c++17 filesystem was the easy route. with c++11 we have to use platform specific code. so different code paths for windows/linux etc.

Reply:
I'll look into this --if there's no gotcha other than different routes for Windows and unix, I can PR to this branch. I'd like just to support as many people as possible 😀

Reply:
I am not targeting merging WITH c++17, but the example disabled by default in cmake.
You(or me) can then patch it later to not use `std::filesystem`.

Reply:
Good idea! We can invest more in this later.

Feel free to mark it as ready for review once you're satisfied.